// oasis-engine/dist/miniprogram.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var miniprogram = require('@oasis-engine/core/dist/miniprogram');
var miniprogram$1 = require('@oasis-engine/loader/dist/miniprogram');
var miniprogram$2 = require('@oasis-engine/math/dist/miniprogram');
var miniprogram$3 = require('@oasis-engine/rhi-webgl/dist/miniprogram');

miniprogram$1.Parser.registerComponents("o3", {
  GLTFModel: miniprogram$1.GLTFModel,
  SpriteRenderer: miniprogram.SpriteRenderer,
  SpriteMask: miniprogram.SpriteMask,
  PointLight: miniprogram.PointLight,
  AmbientLight: miniprogram.AmbientLight,
  DirectLight: miniprogram.DirectLight,
  ParticleRenderer: miniprogram.ParticleRenderer,
  BoxCollider: miniprogram.BoxCollider,
  Camera: miniprogram.Camera,
  Model: miniprogram$1.Model,
  Component: miniprogram.Component,
  SphereCollider: miniprogram.SphereCollider,
  Animator: miniprogram.Animator
}); //@ts-ignore

var version = "0.5.7";
console.log("oasis engine version: " + version);

exports.version = version;
Object.keys(miniprogram).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return miniprogram[k]; }
  });
});
Object.keys(miniprogram$1).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return miniprogram$1[k]; }
  });
});
Object.keys(miniprogram$2).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return miniprogram$2[k]; }
  });
});
Object.keys(miniprogram$3).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return miniprogram$3[k]; }
  });
});
